### ISSUE TITLE
torch.sparse.sum result has wrong dtype when reducing over all dimensions

### ISSUE THREAD

**Issue by pearu** [pytorch/pytorch#65392]
## 🐛 Bug

As in the title.

## To Reproduce

```python
>>> import torch
>>> t = torch.sparse_coo_tensor([[0, 1], [1, 0]], [1, 2], dtype=torch.int16)
>>> torch.sparse.sum(t).dtype
torch.int64
```

## Expected behavior

According to https://pytorch.org/docs/master/generated/torch.sparse.sum.html#torch.sparse.sum, the default dtype of result is the dtype of input:
```python
>>> torch.sparse.sum(t).dtype
torch.int16
```

## Environment

 - PyTorch Version (e.g., 1.0): 1.10.0a0

## Additional context

When `dim` argument is specified, the result dtype is correct:
```python
>>> torch.sparse.sum(t, dim=0).dtype
torch.int16
```
but
```python
>>> torch.sparse.sum(t, dim=(0, 1)).dtype
torch.int64
```
is still incorrect.

Moreover, the explicit usage of dtype argument is sometimes ignored:
```
>> torch.sparse.sum(t, dim=(0, 1), dtype=t.dtype).dtype   # expecting torch.int16
torch.int64
>>> torch.sparse.sum(t, dtype=t.dtype).dtype
torch.int16
```

**Comment by soulitzer** [comment]
In the non-sparse case, looks like the result of sum isn't necessarily guaranteed to be the dtype of the input, so maybe that is just a doc issue? 
Ignoring the dtype argument when explicitly specified though seems bad

**Comment by ngimel** [comment]
As @soulitzer pointed out, `sum` for the dense case produces different dtype than input (e.g. int64 for int16 inputs, like in the example above), so it looks that `dim` behavior for sparse is wrong, rather than no-dim behavior. 
The case where explicit dtype is ignored though is really a bug.

**Comment by rok** [comment]
If `dim` covers all dimensions `sum` will return int64 and `input.dtype` if it only covers some dimensions.

Is this acceptable if documented?

**Comment by rok** [comment]
I would naively expect `input.dtype == output.dtype` if `dtype=None` However maybe there's reasons to be more relaxed that I am not aware of (e.g. type promotion to avoid overflow).

Would this behavior be acceptable?
```
dtype = torch.int16
x = torch.sparse_coo_tensor([[0, 1], [1, 0]], [1, 2], dtype=dtype, device=device)

self.assertEqual(torch.sparse.sum(x).dtype, torch.int64)
self.assertEqual(torch.sparse.sum(x, dim=0).dtype, dtype)
self.assertEqual(torch.sparse.sum(x, dim=(0, 1)).dtype, torch.int64)

self.assertEqual(torch.sparse.sum(x, dtype=dtype).dtype, dtype)
self.assertEqual(torch.sparse.sum(x, dim=0, dtype=dtype).dtype, dtype)
self.assertEqual(torch.sparse.sum(x, dim=(0, 1), dtype=dtype).dtype, dtype)
```

**Comment by rok** [comment]
Agreed with making the behavior more predictable, so how about:

```
dtype = torch.int16
x = torch.sparse_coo_tensor([[0, 1], [1, 0]], [1, 2], dtype=dtype, device=device)

self.assertEqual(torch.sparse.sum(x).dtype, dtype)
self.assertEqual(torch.sparse.sum(x, dim=0).dtype, dtype)
self.assertEqual(torch.sparse.sum(x, dim=(0, 1)).dtype, dtype)

self.assertEqual(torch.sparse.sum(x, dtype=dtype).dtype, dtype)
self.assertEqual(torch.sparse.sum(x, dim=0, dtype=dtype).dtype, dtype)
self.assertEqual(torch.sparse.sum(x, dim=(0, 1), dtype=dtype).dtype, dtype)
```
